text/2978-stack_based_vec.md

```diff
@@ -86,22 +86,26 @@ Of course, fixed buffers lead to some inflexibility because unlike `Vec`, the un
 # Reference-level explanation
 [reference-level-explanation]: #reference-level-explanation
 
-`ArrayVec` is a contiguous memory block where elements can be collected, threfore, a collection by definition and even though `core::collections` doesn't exist, it is the most natural module placement.
+`ArrayVec` is a contiguous memory block where elements can be collected, therefore, a collection by definition and even though `core::collections` doesn't exist, it is the most natural module placement.
 
-The API basically mimics most of the current `Vec` surface with some tweaks and additional methods to manage capacity. Notably, these additional methods are fallible versions of some well-known functions like `insert` that will return `Result` instead of panicking at run-time.
+The API basically mimics most of the current `Vec` surface with some tweaks to manage capacity. Notably, these additional methods are fallible versions of some well-known functions like `push` that will return `Result` instead of panicking at run-time.
 
 ```rust
 // core::collections
 
 pub struct ArrayVec<T, const N: usize> {
     data: MaybeUninit<[T; N]>,
-    len: usize
+    len: usize,
 }
 
 impl<T, const N: usize> ArrayVec<T, N> {
     // Constructors
 
-    pub unsafe fn from_raw_parts(_ptr: *mut T, _len: usize) -> Self;
+    #[inline]
+    pub fn from_array(array: [T; N]) -> Self;
```

Comment:
Isn't this just `ArrayVec::from(arr)`, I don't feel like a method is needed here.

Reply:
Do or will eventually const generics allow something like `impl<T, const N: usize, const M: usize> impl From<[T; M]> for ArrayVec<T, N> where {M <= N}`?

Reply:
yes, using `feature(const_evaluatable_checked)`. This does require a `N <= N` where bound when using it in a generic context though We probably will be able to detect that `Param <= Param` always holds in the future, but for now that is not a priority.

Note that this from impl will be horrible for type inference, so we probably want to make it a separate method anyways imo.

Reply:
@icnr, when you said:

> Isn't this just `ArrayVec::from(arr)`, I don't feel like a method is needed here.

I personally found with [staticvec](https://github.com/slightlyoutofphase/staticvec) that having three inherent methods, (`new_from_slice`, `new_from_array`, and `new_from_const_array`) allowed me to really easily write no less than eight slice / array related `From` impls in direct terms of just calling the appropriate one of those three:

```rust
impl<T, const N: usize> From<[T; N]> for StaticVec<T, N>
impl<T, const N1: usize, const N2: usize> From<[T; N1]> for StaticVec<T, N2>
impl<T: Copy, const N: usize> From<&'_ [T; N]> for StaticVec<T, N>
impl<T: Copy, const N1: usize, const N2: usize> From<&'_ [T; N1]> for StaticVec<T, N2>
impl<T: Copy, const N: usize> From<&'_ [T]> for StaticVec<T, N>
impl<T: Copy, const N: usize> From<&'_ mut [T; N]> for StaticVec<T, N>
impl<T: Copy, const N1: usize, const N2: usize> From<&'_ mut [T; N1]> for StaticVec<T, N2>
impl<T: Copy, const N: usize> From<&'_ mut [T]> for StaticVec<T, N>
```

I was additionally able to implement my `staticvec!` macro in direct terms of `new_from_const_array`, so there was even more code re-use benefits there.

You can see how the three inherent methods work [starting here](https://github.com/slightlyoutofphase/staticvec/blob/bf3b11800cd1ccad176b9538c5a6e4fa4ee98299/src/lib.rs#L156) if you like.